<?xml version="1.0" encoding="utf-8"?>
<robot name="snakebot_v1">
  
  <material name="Blue">
    <color rgba="0.2 0.1 0.8 1.0"/>
  </material>
  <material name="Green">
    <color rgba="0.22 0.32 0.14 1.0"/>
  </material>

  <!-- attach the base_link to world -->
  <link name="world"/>

  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
  </joint>

  <link name="base_link">
    <visual>
      <origin
        xyz="0. 0. 0."
        rpy="1.57079632679 0 1.57079632679"
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/bridge.stl" scale="0.0013 0.0013 0.0013"/>
      </geometry>
      <material name="Blue"/>
    </visual>
    
    <!-- in gazebo, there has to be a mass and an inertia for each link -->
    <inertial>
      <origin xyz="0 0 0.036" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
    ixx="0.1" ixy="0.0" ixz="0.0"
    iyy="0.1" iyz="0.0"
    izz="0.1"/>
    </inertial>
  </link>

  <joint name="rod_one_joint" type="continuous">
    <origin
      xyz="0. 0. 0."
      rpy="0. 0. 0."
    />
    <parent link="base_link"/>
    <child link="link2"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.5" soft_upper_limit="1.5"/>
    <limit effort="0.01" lower="-1" upper="1" velocity="0.5"/>
    <dynamics damping="0.7"/>
  </joint>

  <link name="link2">
    <visual>
      <origin
        xyz="0. 0. 0.072"
        rpy="1.57079632679 0 0."
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/bridge.stl" scale="0.0013 0.0013 0.0013"/>
      </geometry>
      <material name="Green"/>
    </visual>

    <inertial>
      <origin xyz="0 0 0.036" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
      ixx="0.1" ixy="0.0" ixz="0.0"
      iyy="0.1" iyz="0.0"
      izz="0.1"/>
      </inertial>
  </link>

  <joint name="rod_two_joint" type="continuous">
    <origin
      xyz="0. 0. 0.072"
      rpy="0. 0. 1.57079632679"
    />
    <parent link="link2"/>
    <child link="link3"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.5" soft_upper_limit="1.5"/>
    <limit effort="0.01" lower="-1" upper="1" velocity="0.5"/>
    <dynamics damping="0.7"/>
  </joint>

  <link name="link3">
    <visual>
      <origin
        xyz="0. 0. 0.072"
        rpy="1.57079632679 0 1.57079632679"
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/bridge.stl" scale="0.0013 0.0013 0.0013"/>
      </geometry>
      <material name="Green"/>
    </visual>

    <inertial>
      <origin xyz="0 0 0.036" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
      ixx="0.1" ixy="0.0" ixz="0.0"
      iyy="0.1" iyz="0.0"
      izz="0.1"/>
    </inertial>
  </link>

  <joint name="rod_three_joint" type="continuous">
    <origin
      xyz="0. 0. 0.072"
      rpy="0. 0. -1.57079632679"
    />
    <parent link="link3"/>
    <child link="link4"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.5" soft_upper_limit="1.5"/>
    <limit effort="0.01" lower="-1" upper="1" velocity="0.5"/>
    <dynamics damping="0.7"/>
  </joint>

  <link name="link4">
    <visual>
      <origin
        xyz="0. 0. 0.072"
        rpy="1.57079632679 0 0."
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/bridge.stl" scale="0.0013 0.0013 0.0013"/>
      </geometry>
      <material name="Green"/>
    </visual>

    <inertial>
      <origin xyz="0 0 0.036" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
      ixx="0.1" ixy="0.0" ixz="0.0"
      iyy="0.1" iyz="0.0"
      izz="0.1"/>
    </inertial>
  </link>

  <joint name="rod_four_joint" type="continuous">
    <origin
      xyz="0. 0. 0.072"
      rpy="0. 0. 1.57079632679"
    />
    <parent link="link4"/>
    <child link="link5"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.5" soft_upper_limit="1.5"/>
    <limit effort="0.01" lower="-1" upper="1" velocity="0.5"/>
    <dynamics damping="0.7"/>
  </joint>

  <link name="link5">
    <visual>
      <origin
        xyz="0. 0. 0.072"
        rpy="1.57079632679 0 1.57079632679"
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/bridge.stl" scale="0.0013 0.0013 0.0013"/>
      </geometry>
      <material name="Green"/>
    </visual>

    <inertial>
      <origin xyz="0 0 0.036" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
      ixx="0.1" ixy="0.0" ixz="0.0"
      iyy="0.1" iyz="0.0"
      izz="0.1"/>
    </inertial>
  </link>

  <joint name="rod_five_joint" type="continuous">
    <origin
      xyz="0. 0. 0.072"
      rpy="0. 0. -1.57079632679"
    />
    <parent link="link5"/>
    <child link="link6"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.5" soft_upper_limit="1.5"/>
    <limit effort="0.01" lower="-1" upper="1" velocity="0.5"/>
    <dynamics damping="0.7"/>
  </joint>

  <link name="link6">
    <visual>
      <origin
        xyz="0. 0. 0.072"
        rpy="1.57079632679 0 0."
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/bridge.stl" scale="0.0013 0.0013 0.0013"/>
      </geometry>
      <material name="Green"/>
    </visual>

    <inertial>
      <origin xyz="0 0 0.036" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
      ixx="0.1" ixy="0.0" ixz="0.0"
      iyy="0.1" iyz="0.0"
      izz="0.1"/>
    </inertial>
  </link>

  <joint name="rod_six_joint" type="continuous">
    <origin
      xyz="0. 0. 0.072"
      rpy="0. 0. 1.57079632679"
    />
    <parent link="link6"/>
    <child link="headlink"/>
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.5" soft_upper_limit="1.5"/>
    <limit effort="0.01" lower="-1" upper="1" velocity="0.5"/>
    <dynamics damping="0.7"/>
  </joint>


  <link name="headlink">
    <visual>
      <origin
        xyz="0.00 -0.009 0.01"
        rpy="0 0 0"
      />
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/head.stl" scale="0.008 0.012 0.008"/>
      </geometry>
      <material name="Green"/>
    </visual>
    <collision>
      <origin
        xyz="0.00 -0.009 0.01"
        rpy="0 0 0"/>
      <geometry>
        <mesh
          filename="package://snakebot_v1/meshes/head.stl" scale="0.008 0.012 0.008"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <mass value="1"/>
      <inertia
      ixx="0.1" ixy="0.0" ixz="0.0"
      iyy="0.1" iyz="0.0"
      izz="0.05"/>
    </inertial>
  </link>
</robot>